RedisLockRegistry, add automatically deleted for redisLock

Meteorkor opened the pull request:
Resolves #3655

Currently, if you don't manually call `expireUnusedOlderThan()`, OutOfMemory can happen.


Therefore, I changed the code so that if the cache exceeds a setting capacity, it is automatically deleted.

default of capacity occupies about 25 MB.
(`locks` size is 1_000_000)


When reading, multiple threads can access it through readLock.
It is controlled through writeLock only when writing.

I added a queue .
Because it is effective in removing the oldest items

Meteorkor committed:
add automatically deleted for redisLock

Meteorkor committed:
add automatically deleted for redisLock

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -92,6 +97,10 @@
 
 	private final Map<String, RedisLock> locks = new ConcurrentHashMap<>();
 
+	private final ConcurrentLinkedDeque<String> queue = new ConcurrentLinkedDeque<>();
```
I don't think we need to worry about `put` and `remove` optimization.
Those `obtain()` and `expireUnusedOlderThan()` are not designed to be used in the heavy concurrent environment.
The first one is just for lock reference, which you can store for reuse.
The second one indeed has to be called on from an exclusive thread.

Probably for a solid logic it is enough to make those to method as `synchronized`, if you don't want to wrap `LinkedHashMap` into a `Collections.synchronizedMap()`.

And we might need to unit-test to cover this new functionality.

Meteorkor replied:
Thank you for your review.
Fixed to use `LinkedHashMap`.
and
Added test case.

Meteorkor committed:
change LinkedHashMap, add tc

artembilan reviewed (changes requested):
Please, consider to run `gradlew clean :spring-integration-redis:testAll` to be sure that all tests pass and Checkstyle rules are not violated.

Thank you for your effort so far!

artembilan reviewed (changes requested):
Please, also consider to add some note into the `redis.adoc` about that new `capacity` option starting with version `5.5.6`.

Meteorkor committed:
checkStyle fix

Fixed to be sorted when lock() is called.
For removeEldestEntry() and expireUnusedOlderThan()

Add setter for capacity and remove ctor

Meteorkor committed:
warning fix spring-integration-redis:testAll

Meteorkor commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -156,17 +180,26 @@ public void setExecutor(Executor executor) {
+				if (now - lock.getLockedAt() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				} else {
+					break;
```
> I don't think this is correct assumption to stop iterating the cache because we don't match that `getLockedAt()` any more. The cache is only for instances to not overhead with them for the same key, but any obtained `Lock` can be locked later in the logic at arbitrary time. Therefore the oldest entry in the cache might be the most fresh with its `lock()` call. So, early exit from such a loop is misleading and may leave really old not locked entries in the cache.

Thanks for the review.

Fixed to be sorted when lock() is called.
For removeEldestEntry() and expireUnusedOlderThan()

Is it ok to change like this?

artembilan replied:
It is not. 
See my comment on the `obtainLock()` below.

Meteorkor commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -133,12 +139,30 @@ public RedisLockRegistry(RedisConnectionFactory connectionFactory, String regist
+	 * @param expireAfter The expiration in milliseconds.
+	 * @param capacity The capacity of cached lock
+	 */
+	public RedisLockRegistry(RedisConnectionFactory connectionFactory, String registryKey, long expireAfter, long capacity) {
```
> I think this new `capacity` option could be just a setter. No need to be precise via ctor.

Add setter for capacity and remove ctor

Meteorkor committed:
edit redis.adoc

* support automatically clean up cache
* RedisLockRegistry.capacity desc

Meteorkor commented:
> Please, also consider to add some note into the `redis.adoc` about that new `capacity` option starting with version `5.5.6`.

edit redis.adoc

Please tell me if there is anything strange

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -287,7 +333,11 @@ private boolean obtainLock() {
 			if (result) {
-				this.lockedAt = System.currentTimeMillis();
+				final long now = System.currentTimeMillis();
+				synchronized (RedisLockRegistry.this.locks) {
```
No, this is not OK.
Better to have a heavy `expireUnusedOlderThan()` iterating over all the entries, than have a block on every single `Lock.lock()`.

Meteorkor replied:
If it is not sorted when `RedisLockRegistry.RedisLock.obtainLock()` is called, the Lock object that was recently called `RedisLockRegistry.RedisLock.lock()` in `removeEldestEntry()` may be removed.

Is this ok because the cache has been removed?

artembilan replied:
I think it is not a big deal, since the `Lock` object is still valid.
We only miss it in the cache for the next `obtain()` call.
I think more like this cache feature should go at all, but it won't heart to have the logic like we do right now: it is very rear situation that some project would have more then 1,000,000 locks in its logic.

Meteorkor replied:
thanks for your comment
Unfortunately, I recently discovered that it is used in a similar case in another project.
It is being used to prevent simultaneous execution of a specific request by a user.

rolled back `obtainLock()`

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -883,3 +883,6 @@ However, the resources protected by such a lock may have been compromised, so su
 
 Starting with version 5.0, the `RedisLockRegistry` implements `ExpirableLockRegistry`, which removes locks last acquired more than `age` ago and that are not currently locked.
+
+String with version 5.5.6, the `RedisLockRegistry` is support automatically clean up cache for redisLocks in `RedisLockRegistry.locks` via `RedisLockRegistry.capacity`.
```
```suggestion
String with version 5.5.6, the `RedisLockRegistry` is support automatically clean up cache for redisLocks in `RedisLockRegistry.locks` via `RedisLockRegistry.setCaapacity()`.
```

Meteorkor replied:
Thanks for the review.
edit redis.adoc

artembilan commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -883,3 +883,6 @@ However, the resources protected by such a lock may have been compromised, so su
 Starting with version 5.0, the `RedisLockRegistry` implements `ExpirableLockRegistry`, which removes locks last acquired more than `age` ago and that are not currently locked.
+
+String with version 5.5.6, the `RedisLockRegistry` is support automatically clean up cache for redisLocks in `RedisLockRegistry.locks` via `RedisLockRegistry.capacity`.
+`capacity` field can be modified via setters, The default is 1,000,000.
```
```suggestion
See its JavaDocs for more information.
```

Meteorkor replied:
Thanks for the review.
edit redis.adoc

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -152,21 +164,52 @@ public void setExecutor(Executor executor) {
 
+	/**
+	 * Rebalance the capacity of cached locks.
+	 * @param capacity The capacity of cached lock
```
Mention, please, here in JavaDoc a default value.

Meteorkor replied:
Thanks for the review.
edit JavaDoc

artembilan commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java:
```diff
@@ -152,21 +164,52 @@ public void setExecutor(Executor executor) {
+				Iterator<Entry<String, RedisLock>> iterator = this.locks.entrySet().iterator();
+				for (int i = 0; i < removeCnt; i++) {
+					iterator.next();
+					iterator.remove();
```
I don't think that someone is going to modify the `capacity` at runtime.
Move over we consult with it in that `removeEldestEntry` at runtime any way.
So, new value is going to be an effect and therefore obsolete entries are going to be removed on the next `put()`.

Meteorkor replied:
Thanks for the review.
Fixed not to rebalance in setter

Meteorkor committed:
edit redis.adoc, RedisLockRegistry

artembilan reviewed (approved):
LGTM.

Pulling locally for final review and merge process...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/e51513230afd0ddad403d8f0ab093c3d7ee9eb59.

@Meteorkor ,

thank you very much for the contribution; looking forward for more!

For example our `JdbcLockRegistry` suffers from the same issue 😉

Meteorkor commented:
@artembilan 

Thanks for the review
Thank you for the opportunity to contribute.

Let's take a further look at the `JdbcLockRegistry`
